File: src/examples/ds1302.py
#!/usr/bin/env python

# DS1302 DataSheet: https://www.analog.com/media/en/technical-documentation/data-sheets/DS1302.pdf

from mpsse import *
from datetime import datetime
from time import sleep


def to_bcd(val):
    val %= 100
    return ((val // 10) << 4) + (val % 10)


def init_time_cmd(cmd):

    now = datetime.now()

    regs = bytes([cmd,
                  to_bcd(now.second),
                  to_bcd(now.minute),
                  to_bcd(now.hour),
                  to_bcd(now.day),
                  to_bcd(now.month),
                  (now.isoweekday() % 7) + 1,
                  to_bcd(now.year),
                  0])
    return regs


def print_bytes(*args):
    out = []
    for arg in args:
        out += [b for b in arg]

    formated = "".join([f"{b:02x} " for b in out])
    print(formated)


def main():
    try:

        ds1302 = MPSSE(SPI3, ONE_MHZ, LSB)

        # The DS1302's chip select pin idles low
        ds1302.SetCSIdle(0)

        print("%s initialized at %dHz (SPI mode 3), LSB and CS low on idle." % (ds1302.GetDescription(), ds1302.GetClock()));

        print("Reading control register.")

        # Get the current control register value - one-byte write command
        ds1302.Start()
        ds1302.Write(bytes([0x8F]))
        control = ds1302.Read(1)[0]
        ds1302.Stop()

        print("Writing control register.")

        # single-byte write command
        # disabling write-protect bit is not strictly necessary in this example
        # because it protects RAM registers only, and we are writing Clock registers
        control &= ~0x80

        # Write the new control value to the control register
        ds1302.Start()
        ds1302.Write(bytes([0x8E]))
        ds1302.Write(bytes([control]))
        ds1302.Stop()

        print("Initializing time:\n");

        # Write the initial date
        init = init_time_cmd(0xBF)
        ds1302.Start()
        ds1302.Write(init)
        ds1302.Stop()

        print_bytes(init[1:])

        print("Reading time in one second intervals (stop with Ctrl+C):");

        # Loop to print the time elapsed every second
        while True:

            try:
                sleep(1)

                # Read in the elapsed seconds
                ds1302.Start()
                ds1302.Write(bytes([0xBF]))
                data = ds1302.Read(8)
                ds1302.Stop()

                print_bytes(data)

            except KeyboardInterrupt:
                break

        ds1302.Close()
        print("Exiting...")

    except Exception as e:
        print("Error reading from DS1302:", e)


if __name__ == "__main__":
    main()

File: src/examples/ds1302.c
/* DS1302 DataSheet: https://www.analog.com/media/en/technical-documentation/data-sheets/DS1302.pdf */

#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <mpsse.h>
#include <time.h>
#include <unistd.h>


/* Timer registers */
typedef struct {
	uint8_t cmd;		/* command byte */
	uint8_t	sec;		/* reg 0 */
	uint8_t	min;		/* reg 1 */
	uint8_t	hour;		/* reg 2 */
	uint8_t	date;		/* reg 4 */
	uint8_t	mon;		/* reg 5 */
	uint8_t	wday;		/* reg 3 */
	uint8_t	year;		/* reg 6 */
	uint8_t ctrl;		/* ret 7 */
} t_ds1302regs;


static const t_ds1302regs set_current_date(void);
volatile uint8_t run = 1;

/* Signal handler function */
static void handle_sigint(int sig) {
	printf("Caught signal %d (Ctrl+C).\n", sig);
	run = 0;
}


int main(void)
{
	/* Register the signal handler for SIGINT */
	signal(SIGINT, handle_sigint);

	struct mpsse_context *ds1302 = NULL;

	int retval = EXIT_FAILURE;

	/* Initialzie DS1302 in SPI3 mode, 1MHz speed and LSB bit order, with CS to low on idle */
	if((ds1302 = MPSSE(SPI3, ONE_MHZ, LSB)) != NULL && ds1302->open)
	{
		SetCSIdle(ds1302, 0);

		printf("%s initialized at %dHz (SPI mode 3), LSB and CS low on idle.\n", GetDescription(ds1302), GetClock(ds1302));
		printf("VID 0x%04x PID 0x%04x\n", GetVid(ds1302), GetPid(ds1302));

		printf("Reading control register.\n");

		char *control = NULL;
		char cmd = 0x8F;

		/* singler-byte read command */
		Start(ds1302);
		Write(ds1302, &cmd, 1);
		control = Read(ds1302, 1);
		Stop(ds1302);

		printf("Writing control register.\n");
		control[0] &= ~0x80;

		/*
		 * single-byte write command
		 * disabling write-protect bit is not strictly necessary in this example
		 * because it protects RAM registers only, and we are writing Clock registers
		 */
		cmd = 0x8E;
		Start(ds1302);
		Write(ds1302, &cmd, 1);
		Write(ds1302, control, 1);
		Stop(ds1302);

		free(control);

		printf("Initializing time:\n");

		t_ds1302regs ds1302regs = set_current_date();

		char *data = (char *)&ds1302regs;
		for (int i = 1; i < sizeof(ds1302regs); ++i)
		{
			printf("%02x ", (uint8_t)data[i]);
		}

		/* clock brust-mode write command */
		ds1302regs.cmd = 0xBE;
		Start(ds1302);
		Write(ds1302, (char*)&ds1302regs, sizeof(ds1302regs));
		Stop(ds1302);

		printf("\n\nReading time in one second intervals (stop with Ctrl+C):\n");

		while(run)
		{
			sleep(1);

			/* clock brust-mode read command */
			cmd = 0xBF;
			Start(ds1302);
			Write(ds1302, &cmd, 1);
			char *data = Read(ds1302, 8);
			Stop(ds1302);

			for (int i = 0; i < 8; ++i)
			{
				printf("%02x ", (uint8_t)data[i]);
			}
			printf("\n");

			free(data);
		}	
		printf("Exiting...\n");
	}
	else
	{
		printf("Failed to initialize MPSSE: %s\n", ErrorString(ds1302));
	}

	Close(ds1302);

	return retval;
}


static inline uint8_t int8_t_to_bcd(int8_t val)
{
	uint8_t v = ((uint8_t) val) % 100;
	return ((v / 10) << 4) + (v % 10); 
}


static inline uint8_t int16_t_to_bcd(int16_t val)
{
	uint8_t v = ((uint16_t) val) % 100;
	return ((v / 10) << 4) + (v % 10); 
}


static const t_ds1302regs set_current_date(void)
{
	time_t t;
	struct tm tm;

	t_ds1302regs regs;
	memset(&regs, 0, sizeof(regs));
	/* repeat if second has changed while converting - try to be accurate within a second */
	do {
		t = time(NULL);
		localtime_r(&t, &tm);

		regs.sec  = int8_t_to_bcd(tm.tm_sec);
		regs.min  = int8_t_to_bcd(tm.tm_min);
		regs.hour = int8_t_to_bcd(tm.tm_hour);
		regs.date = int8_t_to_bcd(tm.tm_mday);
		regs.mon  = int8_t_to_bcd(tm.tm_mon+1);
		regs.wday = tm.tm_wday + 1;
		regs.year = int16_t_to_bcd(tm.tm_year);
	} while (t != time(NULL));
	return regs;
}

